### PR TITLE
Paginate template options

### PR DESCRIPTION
**Pull request by aq17**
Addresses #10078 

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/cmd/pulumi/new.go [hunk 1]
```diff
@@ -776,7 +776,7 @@ func chooseTemplate(templates []workspace.Template, opts display.Options) (works
 		if err := survey.AskOne(&survey.Select{
 			Message:  message,
 			Options:  options,
-			PageSize: len(options),
+			PageSize: 10,
```
Comment (AaronFriel) [inline pkg/cmd/pulumi/new.go:779]: That's surprisingly easy.

**Comment by AaronFriel** [comment]
@aq17 going to expand on this to ask some UX questions, perhaps you could find time with someone on @pulumi/user-experience team to go over options?

Some of these are "current day" problems, but a bit exacerbated by only showing 10 items.

I don't see an affordance that there are more templates after the 10th item:
image

I have to select "Show additional templates" to see all, and _then_ I see a list of additional options and the prompt appears... twice?

image

And lastly, without selecting "Show additional templates", filtering only operates on the present list:

image

**Comment by iwahbe** [comment]
If I understand correctly, this should prevent us from needing to use "Show additional templates", since we can just show all templates. The user can just scroll down to see the rest.

**Comment by aq17** [comment]
> If I understand correctly, this should prevent us from needing to use "Show additional templates", since we can just show all templates. The user can just scroll down to see the rest.

Great point, also an easy change 😄

**Comment by dixler** [comment]
@AaronFriel 
> I have to select "Show additional templates" to see all, and _then_ I see a list of additional options and the prompt appears... twice?
> 
> 
> 
> image

https://github.com/pulumi/pulumi/issues/8169

I fixed this before by **upgrading survey by a major version** and it got reverted.

**Comment by aq17** [comment]
> @AaronFriel
> 
> > I have to select "Show additional templates" to see all, and _then_ I see a list of additional options and the prompt appears... twice?
> > image
> 
> #8169
> 
> I fixed this before by **upgrading survey by a major version** and it got reverted because of major bugs

Might be worth upgrading to v2 everywhere we use the package, it doesn't look like there are too many breaking changes

File: pkg/cmd/pulumi/new.go [hunk 2]
```diff
@@ -756,7 +756,7 @@ func chooseTemplate(templates []workspace.Template, opts display.Options) (works
 	message := "\rPlease choose a template:"
 	message = opts.Color.Colorize(colors.SpecPrompt + message + colors.Reset)
 
-	showAll := false
+	showAll := true
```
Comment (iwahbe) [inline pkg/cmd/pulumi/new.go:759]: We only set `showAll` to `true`, so if it starts as `true`, we can just omit the variable.

**Comment by dixler** [comment]
@aq17 can we list the number of templates that are on the list inside of the prompt? A user might not realize that the list scrolls.

File: pkg/cmd/pulumi/policy_new.go [hunk 3]
```diff
@@ -297,7 +297,7 @@ func choosePolicyPackTemplate(templates []workspace.PolicyPackTemplate,
 	if err := survey.AskOne(&survey.Select{
 		Message:  message,
 		Options:  options,
-		PageSize: len(options),
+		PageSize: 10,
```
Comment (dixler) [inline pkg/cmd/pulumi/policy_new.go:300]: we should use the height of the user's terminal to determine the size of the page

it looks good inside vscode, but leaves a bit to be desired in a full sized terminal.
2022-07-15-165446_2560x1051_scrot
.

**Review by dixler — CHANGES_REQUESTED** [review]

File: pkg/cmd/pulumi/policy_new.go (hunk 3 above)
Comment (aq17) [reply pkg/cmd/pulumi/policy_new.go:300]: Just tested it, definitely looks better but there's the issue with the `Please choose a template (#/#) shown:` prompt appearing numerous times as you scroll again

Comment (aq17) [reply pkg/cmd/pulumi/policy_new.go:300]: Actually, it looks like providing a small buffer (terminal height - 5) solves that, but the flickering is 😠 
Screen Shot 2022-07-15 at 5 07 24 PM

Comment (AaronFriel) [reply pkg/cmd/pulumi/policy_new.go]: I would be inclined to set a lower limit to reduce flickering. Is there an upstream issue to track that flickering?

Comment (AaronFriel) [reply pkg/cmd/pulumi/policy_new.go]: I meant with the CLI library - is this something other folks have noticed or the upstream repo is aware of? The flickering is very noticeable - even more-so on Windows - and I think it'd be good to flag that with the maintainer.

Comment (aq17) [reply pkg/cmd/pulumi/policy_new.go]: Ah no, it's also still there even after upgrading to their v2; I'll open an issue now.

File: pkg/cmd/pulumi/new.go [hunk 4]
```diff
@@ -768,6 +769,8 @@ func chooseTemplate(templates []workspace.Template, opts display.Options) (works
 		if height > len(options) {
 			height = len(options)
 		}
+
+		height = height - buffer
```
Comment (AaronFriel) [inline pkg/cmd/pulumi/new.go:794]: If we change this height to a much smaller value, is the flickering as severe?

Comment (aq17) [reply pkg/cmd/pulumi/new.go:794]: Don't think there's a clear maximum, even something as small as 10 on a max sized terminal causes pretty bad flickering on my mac default terminal

**Review by dixler — APPROVED** [review]
I think this PR does resolve the issue that it's addressing. We should :ship: it since we have another issue that tracks the flickering.

**Comment by aq17** [comment]
@dixler [upstream issue](https://github.com/AlecAivazis/survey/issues/436#issuecomment-1203111689) had a few other people commenting on ideas regarding the flickering  – can we still merge this and address flickering in a separate PR?

**Pull request merged**